[None][feat] Add NCCL device kernels for AR+RMS fusion

nv-lschneider opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Added AllReduce strategy NCCL_DEVICE with device-side fused AllReduce (residual RMSNorm), multimem support, and launch-configuration caching; exposed in Python and runtime config; userbuffers updated to support it.

- **Tests**
  - Added multi-GPU NCCL device tests and extended microbenchmark with only-ub option and improved size reporting.

- **Documentation**
  - API references updated to include NCCL_DEVICE.

- **Chores**
  - Updated spell-check and pre-commit configuration.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

## Description

The MR introduces a new kernel launch mechanism to support kernels with the NCCL device API.
It implements 1 kernel to start with: RESIDUAL_RMS_NORM for fp16 types.

This new kernel is meant to replace/enhance the performance of AllReduce using the stable NCCL API.
This is the first kernel of potentially more variations for best performance. The default AR selection strategy is not impacted yet.

It is designed to be low latency for small to medium message sizes.

The MR uses the existing NCCLUBAllocator and extends it to hold necessary persistent resources like NCCL registered memory windows and device communicators.

The AllReduce Operation is implemented as a new AllReduceStrategy and launched from the AllReduce plugin. `cpp/tensorrt_llm/thop/allreduceOP.cpp`.
It launches its own new kernels at `cpp/tensorrt_llm/kernels/nccl_device`.

The kernel itself is highly templated to be flexible for future demands without impeding runtime performance.

This MR implements the new kernel in a two-shot / fp16 implementation first.
It is already competitive in this form, however after adoption of this kernel further modifications and additions can be included.
1. one-shot flavor
2. fp8 and fp4
 support in the future.


## Test Coverage

 - [x] Python unit test: `tests/unittest/_torch/multi_gpu/test_nccl_device.py`
 - [x] Microbenchmakr: `tests/microbenchmark/all_reduce.py`
 
 The microbenchmark has been updated slightly. It includes now the new strategy and optionally also UB for comparison.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

# Important Caveat

This change requires NCCL 2.28 to run successfully.
SInce the current dev container of TRT-LLM does not use 2.28 yet, I would like to gather some feedback before 2.28 becomes available.
A real test will only be possible when version 2.28 is included in the dev container.

nv-lschneider committed:
Introducing a new AR strategy that makes use if NCCL symmetric memory and NEW NCCL device API to use NCCL to fuse RMS Norm with AllReduce.

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

Superjomn committed:
[https://nvbugs/5351244][fix] CHERRY-PICK test_mpi_session (#7501) (#7900)

Signed-off-by: Yan Chunwei <328693+Superjomn@users.noreply.github.com>
Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

pre-commit changes

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

clang formatting

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

safe guarding NCCL 2.27 build

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

fixing precommit formatting

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

most of code rabbit comments

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

adding missing semi-colon

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

removing unused comment lines

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

Clarifying the test on how to compre residual chunked and unchunked.

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

fixing pre-commit

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

fixing pre-commit

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

fixing missing variable, rebase complete and tested

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

using a grid stride loop with less blocks launched for large message sizes

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

using functioning grid stride loop for NCCL_DEVICE. It helps with better performance at larger message sizes

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

initial oneshot implementation

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

minor tweaks to include one shot

fixes

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

enabling grid stride loop, but no perf benefit.

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

addressing review feedback

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

fix formatting

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

nv-lschneider commented on cpp/tensorrt_llm/kernels/nccl_device/config.cu:
```diff
@@ -0,0 +1,487 @@
+
+    // Delegate all launch logic to the config class
+    this->launchKernel(inWindow, outWindow, residual, residualOutWindow, weight, bias, devComm, eps, stream);
+}
```
No, these are correct.
This is exactly how the launch config sets it up and it mirrors UB.

Tabrizian commented on tests/unittest/_torch/multi_gpu/test_nccl_device.py:
```diff
@@ -0,0 +1,197 @@
+        chunk_start = rank * residual_chunk_size
+        chunk_end = min((rank + 1) * residual_chunk_size, ref_residual.size(0))
+
+        # If we do perform the AllGaterh implicitly we can compare the entire tensor.
```
```suggestion
        # If we do perform the AllGather implicitly we can compare the entire tensor.
```

nv-lschneider commented on cpp/tensorrt_llm/kernels/nccl_device/config.cu:
```diff
@@ -0,0 +1,516 @@
+        return false;
+    }
+
+    // 6. Query actual kernel resource usage from kernel pointer for the specific unroll factor
```
Yes, the numbering was out of order.
I removed the numbering, since it is isn't necessary and brittle change.

nv-lschneider committed:
first step new allreduce

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

better UB init handling

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

accept multiple strategies

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

test to debug mnnvl

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

rebasing and addressing comments

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

remove unneeded type decl

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

Tabrizian commented on cpp/tensorrt_llm/kernels/nccl_device/config.cu:
```diff
@@ -0,0 +1,516 @@
+    {
+        int local_sms = 1;
+        int dev = -1;
+        cudaError_t cudaStatus = cudaGetDevice(&dev);
```
Can we use `TLLM_CUDA_CHECK` instead here and everywhere?
```suggestion
        TLLM_CUDA_CHECK(cudaGetDevice(&dev));
```

Tabrizian commented on cpp/tensorrt_llm/kernels/nccl_device/config.cu:
```diff
@@ -0,0 +1,516 @@
+{
+    // Get CUDA device properties
+    int dev = -1;
+    cudaError_t cudaStatus = cudaGetDevice(&dev);
```
same as above

Tabrizian commented on cpp/tensorrt_llm/thop/allreduceOp.cpp:
```diff
@@ -338,6 +349,102 @@ class AllreduceOp
         return fallbackRunSubsequentOps(input, residual, norm_weight, scale, bias, norm_out);
     }
 
+    std::vector<torch::Tensor> runNCCLAllReduceDeviceFusion(torch::Tensor const& input,
```
Thanks, I agree with having two separate strategies for now. Resolving.

nv-lschneider commented on cpp/tensorrt_llm/kernels/nccl_device/config.h:
```diff
@@ -0,0 +1,180 @@
+ * Copyright (c) 2025, NVIDIA CORPORATION. All rights reserved.
+ *
+ * See LICENSE.txt for license information
+ ************************************************************************/
```
I updated the license of the 2 files. And double checked that the other files are OK too.

nv-lschneider committed:
configurable explore 2d option

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

nv-lschneider committed:
reset files, I don't want to touch

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

nv-lschneider reviewed (commented):
I rebased the code and addressesd your comments. Thx for the patience.
Rebasing takes awhile.

nv-lschneider committed:
fixing precommit

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

nv-lschneider committed:
fix precommit

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

nv-lschneider committed:
printing csv output correctly

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

nv-lschneider committed:
remove auto skip

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

nv-lschneider committed:
instrumenting to investigate potential fallbacks

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

nv-lschneider committed:
fixes

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

nv-lschneider committed:
more instrumenting

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

nv-lschneider committed:
remove old check

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

nv-lschneider committed:
exclude NCCL_DEVICE from stupid checks

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

nv-lschneider committed:
topology detection skip for stratgies that don't need it

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

nv-lschneider committed:
refactor into function

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

nv-lschneider committed:
reverting to simpler strategy

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>

Tabrizian commented on cpp/tensorrt_llm/kernels/nccl_device/kernels.cuh:
```diff
@@ -0,0 +1,257 @@
+/*
```
Sounds good, thanks.

Tabrizian reviewed (commented):
LGTM, merging is blocked until the PyTorch container upgrades to 2.28 version.

litaotju commented on cpp/tensorrt_llm/kernels/nccl_device/multimem.h:
```diff
@@ -0,0 +1,273 @@
+#define TRTLLM_NCCL_DEVICE_MULTIMEM_H
+
+#include <cuda_fp16.h>
+#if CUDART_VERSION >= 11000
```
I think this is not needed, we only support cuda 12.9 and 13 now.

litaotju commented on cpp/tensorrt_llm/kernels/nccl_device/multimem.h:
```diff
@@ -0,0 +1,273 @@
+
+// Base template for multimemLoadSum - device assert for SM < 90
+template <typename ptrT, typename valT>
+__device__ __forceinline__ valT multimemLoadSum(ptrT const* addr)
```
shall this file has suffix ".cuh" since it has `__device__` identifier?

nv-lschneider committed:
Reviewing feedback #1

Signed-off-by: Ludwig Schneider <lschneider@nvidia.com>
